netbox/netbox/models/features.py

```diff
@@ -337,6 +340,87 @@ class Meta:
         abstract = True
 
 
+class RenderMixin(models.Model):
+    """
+    Enables support for rendering templates.
+    """
+    template_code = models.TextField(
+        verbose_name=_('template code'),
+        help_text=_('Jinja2 template code.')
+    )
+    environment_params = models.JSONField(
+        verbose_name=_('environment parameters'),
+        blank=True,
+        null=True,
+        default=dict,
+        help_text=_(
+            'Any <a href="https://jinja.palletsprojects.com/en/3.1.x/api/#jinja2.Environment">additional parameters</a>'
+            ' to pass when constructing the Jinja2 environment.'
+        )
+    )
+    mime_type = models.CharField(
+        max_length=50,
+        blank=True,
+        verbose_name=_('MIME type'),
+        help_text=_('Defaults to <code>text/plain; charset=utf-8</code>')
+    )
+    file_name = models.CharField(
+        max_length=200,
+        blank=True,
+        help_text=_('Filename to give to the rendered export file')
+    )
+    file_extension = models.CharField(
+        verbose_name=_('file extension'),
+        max_length=15,
+        blank=True,
+        help_text=_('Extension to append to the rendered filename')
+    )
+    as_attachment = models.BooleanField(
+        verbose_name=_('as attachment'),
+        default=True,
+        help_text=_("Download file as attachment")
+    )
+
+    class Meta:
+        abstract = True
+
+    def get_context(self, context=None, queryset=None):
+        raise NotImplementedError(_("{class_name} must implement a get_context() method.").format(
+            class_name=self.__class__
+        ))
+
+    def render(self, context=None, queryset=None):
+        """
+        Render the template with the provided context. The context is passed to the Jinja2 environment as a dictionary.
+        """
+        context = self.get_context(context=context, queryset=queryset)
+        env_params = self.environment_params or {}
+        output = render_jinja2(self.template_code, context, env_params)
+
+        # Replace CRLF-style line terminators
+        output = output.replace('\r\n', '\n')
+
+        return output
+
+    def render_to_response(self, context=None, queryset=None):
+        output = self.render(context=context, queryset=queryset)
+        mime_type = self.mime_type or 'text/plain; charset=utf-8'
+
+        # Build the response
+        response = HttpResponse(output, content_type=mime_type)
+
+        if self.as_attachment:
+            extension = f'.{self.file_extension}' if self.file_extension else ''
+            if queryset:
+                filename = self.file_name or filename_from_model(queryset.model)
```

Comment:
Won't `queryset.model` always be this class? We should be able to remove `queryset` as a kwarg for this method and instead reference `self._meta.model`.

Reply:
At this point, `self._meta.model` refers to either `ExportTemplate` or `ConfigTemplate`, not the actual exported/rendered config `Model`. So, to use the device name or device type as the `filename`, we need to rely on the `queryset`.

I also use `queryset` at the beginning of the function to call `self.render`, which internally calls `self.get_context`. Because of this, passing `queryset` through `kwargs` is currently necessary.

While it might be possible to pass both the `context` dict and `queryset` together in a single variable for `render` and `get_context`, doing so could make the function's purpose harder to understand.